src/storage/src/render/mod.rs

```diff
@@ -144,105 +141,116 @@ pub fn build_storage_dataflow<A: Allocate, B: StorageCapture>(
-
-                    let mut tokens = Vec::new();
-                    let ok = source(region, "InvalidSource", |cap, info| {
-                        let mut act_cap = Some(ActivateCapability::new(
-                            cap,
-                            &info.address,
-                            region.activations(),
-                        ));
-
-                        let drop_activator = Arc::new(scope.sync_activator_for(&info.address[..]));
-                        let drop_activator_weak = Arc::downgrade(&drop_activator);
-
-                        tokens.push(drop_activator);
-
-                        move |_handle| {
-                            if drop_activator_weak.upgrade().is_some() {
-                                act_cap.take();
-                            }
+            let ((ok, err), token) = crate::render::sources::render_source(
+                region,
+                &debug_name,
+                &as_of,
+                source.id,
+                source.desc.clone(),
+                source.storage_metadata.clone(),
+                storage_state,
+            );
+            storage_state
+                .source_tokens
+                .insert(source.id, Arc::clone(&token));
+
+            let operator_name = format!("persist_sink({})", source.storage_metadata.persist_shard);
+            let mut persist_op = OperatorBuilder::new(operator_name, region.clone());
+
+            // We want exactly one worker (in the cluster) to send all the data to persist. It's fine
+            // if other workers from replicated clusters write the same data, though. In the real
+            // implementation, we would use a storage client that transparently handles writing to
+            // multiple shards. One shard would then only be written to by one worker but we get
+            // parallelism from the sharding.
+            // TODO(aljoscha): Storage must learn to keep track of collections, which consist of
+            // multiple persist shards. Then we should set it up such that each worker can write to one
+            // shard.
+            let hashed_id = source.id.hashed();
+            let active_write_worker = (hashed_id as usize) % scope.peers() == scope.index();
+
+            let source_data = ok.map(Ok).concat(&err.map(Err)).inner;
+            let mut input = persist_op.new_input(&source_data, Exchange::new(move |_| hashed_id));
+
+            let shared_frontier = Rc::clone(&storage_state.source_uppers[&source.id]);
+
+            let weak_token = Arc::downgrade(&token);
+
+            persist_op.build_async(
+                region.clone(),
+                move |mut capabilities, frontiers, scheduler| async move {
+                    capabilities.clear();
+                    let mut buffer = Vec::new();
+                    let mut stash: HashMap<_, Vec<_>> = HashMap::new();
+
+                    let mut write = crate::persist_cache::open_writer::<
```

Comment:
I put it in the value field because a row isn't necessarily unique, so calling it a key made less sense to me semantically. The potential for future indexed access is a convincing enough argument though so I'll switch them around